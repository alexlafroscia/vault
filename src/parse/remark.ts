import type { Root } from "mdast";
import type { Compatible } from "vfile";
import { unified } from "unified";
import remarkParse from "remark-parse";

import { remarkCallout, makeOptions as makeCalloutOptions } from "./callout.js";
import { type Frontmatter, remarkFrontmatter, matter } from "./frontmatter.js";
import {
    remarkImageTransclusion,
    type RemarkImageTransclusionOptions,
} from "./transclusion.js";
import {
    remarkWikiLink,
    makeOptions as makeRemarkWikiLinkOptions,
    type MakeOptionsOptions as MakeRemarkWikiLinkOptionsOptions,
} from "./wiki-link.js";

export interface ParseResult {
    ast: Root;
    frontmatter: Frontmatter;
}

export interface MakeParserOptions
    extends Partial<MakeRemarkWikiLinkOptionsOptions>,
        Partial<RemarkImageTransclusionOptions> {}

export function makeParser(options: MakeParserOptions = {}) {
    const parser = unified()
        .use(remarkParse)
        .use(remarkCallout, makeCalloutOptions())
        .use(remarkFrontmatter, ["yaml"])
        .use(
            remarkWikiLink,
            makeRemarkWikiLinkOptions({
                // Defaults
                permalinks: [],
                href(permalink) {
                    return { type: "external", permalink };
                },

                // Overrides
                ...options,
            }),
        )
        .use(remarkImageTransclusion, options);

    return function parse(doc: Compatible): ParseResult {
        const parsedAST = parser.parse(doc);
        const transformedAST = parser.runSync(parsedAST, doc);

        return {
            ast: transformedAST,
            frontmatter: matter(transformedAST),
        };
    };
}
