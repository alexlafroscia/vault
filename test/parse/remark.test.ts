import { expect, test } from "vitest";
import dedent from "dedent";

import { makeParser } from "~test/helpers/remark";

test("basic markdown", () => {
    const parse = makeParser();
    const { ast, frontmatter } = parse(dedent`
        # Heading 1

        Paragraph Text
    `);

    expect(ast).toMatchSnapshot();
    expect(frontmatter).toEqual({});
});

test("frontmatter extraction", () => {
    const parse = makeParser();
    const { ast, frontmatter } = parse(dedent`
        ---
        foo: bar
        ---

        # Heading 1
    `);

    expect(ast).toMatchSnapshot();
    expect(frontmatter).toEqual({
        foo: "bar",
    });
});
