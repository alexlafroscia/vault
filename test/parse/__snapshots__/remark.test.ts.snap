// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic markdown 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "Heading 1",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 12,
          "line": 1,
          "offset": 11,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 15,
              "line": 3,
              "offset": 27,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 13,
            },
          },
          "type": "text",
          "value": "Paragraph Text",
        },
      ],
      "position": {
        "end": {
          "column": 15,
          "line": 3,
          "offset": 27,
        },
        "start": {
          "column": 1,
          "line": 3,
          "offset": 13,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 15,
      "line": 3,
      "offset": 27,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`frontmatter extraction 1`] = `
{
  "children": [
    {
      "position": {
        "end": {
          "column": 4,
          "line": 3,
          "offset": 16,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "yaml",
      "value": "foo: bar",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 5,
              "offset": 29,
            },
            "start": {
              "column": 3,
              "line": 5,
              "offset": 20,
            },
          },
          "type": "text",
          "value": "Heading 1",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 12,
          "line": 5,
          "offset": 29,
        },
        "start": {
          "column": 1,
          "line": 5,
          "offset": 18,
        },
      },
      "type": "heading",
    },
  ],
  "position": {
    "end": {
      "column": 12,
      "line": 5,
      "offset": 29,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
